### ISSUE TITLE
Clients View in Dashboard: Request failed with status code 500

### ISSUE THREAD

**Issue by victorclaessen** [emqx/emqx#3668]
#### Environment

- OS: Debian 10
- Erlang/OTP: R22/10.7.1
- EMQ: 4.1.1

#### Description

I have a similar problem as #3248 (which was supposedly solved in 4.0.3) in my version 4.1.1; whenever I go to the Dashboard menu > Monitoring > Clients, and select a specific client, I get a red bar from top, saying Request failed with status code 500.

```
2020-08-14 15:51:07.881 [error] GET /api/v4/clients/clientredacted error: badarg, stacktrace:
[{erlang,'++',
     [{error,
          {'EXIT',
              {{badfun,#Fun<emqx_mgmt_api_clients.3.76335233>},
               [{emqx_mgmt,lookup_client,3,
                    [{file,
                         "/emqx-rel/_build/emqx-pkg/lib/emqx_management/src/emqx_mgmt.erl"},
                     {line,184}]},
                {rpc,'-handle_call_call/6-fun-0-',5,
                    [{file,"rpc.erl"},{line,197}]}]}}},
      [#{port => 41472,created_at => <<"2020-08-14 07:00:48">>,
         recv_cnt => 88199,max_awaiting_rel => 100,recv_oct => 10957376,
         mqueue_len => 0,recv_msg => 95083,
         clientid => <<"clientredacted">>,
         reductions => 102924323,send_pkt => 128362,keepalive => 60,
         max_inflight => 32,node => 'emqx@hostnameredacted',
         send_cnt => 128361,proto_ver => 4,zone => external,
         max_mqueue => 1000,username => undefined,
         connected_at => <<"2020-08-14 07:00:48">>,heap_size => 4185,
         recv_pkt => 126771,connected => true,expiry_interval => 7200,
         is_bridge => false,send_oct => 7428458,mailbox_len => 0,
         inflight => 0,mqueue_dropped => 0,send_msg => 33272,
         max_subscriptions => 0,clean_start => false,awaiting_rel => 0,
         proto_name => <<"MQTT">>,subscriptions_cnt => 6,
         ip_address => <<"192.168.4.25">>}]],
     []},
 {lists,append,1,[{file,"lists.erl"},{line,127}]},
 {emqx_mgmt_api_clients,lookup,2,
     [{file,
          "/emqx-rel/_build/emqx-pkg/lib/emqx_management/src/emqx_mgmt_api_clients.erl"},
      {line,127}]},
 {minirest_handler,dispatch,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/minirest/src/minirest_handler.erl"},
      {line,84}]},
 {minirest,handle_request,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/minirest/src/minirest.erl"},
      {line,112}]},
 {minirest,init,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/minirest/src/minirest.erl"},
      {line,104}]},
 {cowboy_handler,execute,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/cowboy/src/cowboy_handler.erl"},
      {line,41}]},
 {cowboy_stream_h,execute,3,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/cowboy/src/cowboy_stream_h.erl"},
      {line,320}]}]
```

*A description of the issue*

**Comment by spring2maz** [comment]
Hi @victorclaessen 

The line numbers from your crash log do not add up to me.
Do you have a cluster of mixed versions ?
The management API client seems to be at version v4.1.1 ,
and the other node (which failed at handling the lookup rpc call) is running at version v4.0.3?

**Comment by spring2maz** [comment]
Or did this happen during (or shortly after) an hot-patch upgrade ?
If ^ this is the case, I'd assume it's transient.

**Comment by victorclaessen** [comment]
There was one unmatched node in the cluster. I now upgraded all the nodes to 4.1.3. Right now the clients view is ok, but now I get the same error when I try to view the rules.

```
(emqx@node1)1> 2020-08-20 18:01:51.698 [error] GET /api/v4/rules error: function_clause, stacktrace:
[{emqx_rule_engine_api,record_to_map,
     [{rule,<<"rule:dbd7a014">>,
          [<<"t/#">>],
          <<"SELECT\n  payload\nFROM\n  \"t/#\"">>,false,
          [{var,<<"payload">>}],
          [],{},{},
          [{action_instance,<<"data_to_webserver_1587892001185828635">>,
               data_to_webserver,
               #{<<"$resource">> => <<"resource:9bfabdc5">>,
                 <<"payload_tmpl">> => <<"{\"text\": \"${payload}\"}">>}},
           {action_instance,<<"data_to_webserver_1587892001187101398">>,
               data_to_webserver,
               #{<<"$resource">> => <<"resource:222b7366">>,
                 <<"payload_tmpl">> => <<"{\"text\": \"${payload}\"}">>}}],
          true,<<>>}],
     [{file,
          "/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine_api.erl"},
      {line,407}]},
 {emqx_rule_engine_api,maybe_record_to_map,2,
     [{file,
          "/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine_api.erl"},
      {line,394}]},
 {lists,foldr,3,[{file,"lists.erl"},{line,1276}]},
 {emqx_rule_engine_api,return_all,1,
     [{file,
          "/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine_api.erl"},
      {line,390}]},
 {minirest_handler,dispatch,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/minirest/src/minirest_handler.erl"},
      {line,84}]},
 {minirest,handle_request,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/minirest/src/minirest.erl"},
      {line,112}]},
 {minirest,init,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/minirest/src/minirest.erl"},
      {line,104}]},
 {cowboy_handler,execute,2,
     [{file,"/emqx-rel/_build/emqx-pkg/lib/cowboy/src/cowboy_handler.erl"},
      {line,41}]}]
```

**Comment by spring2maz** [comment]
@victorclaessen This is unfortunately a bug. 

@terry-xiaoyu @HJianBo do we have an upgrade support for saved rules ?
it seems the a few records have extended with new fields if you diff `v4.0.3...v4.1.3`

**Comment by victorclaessen** [comment]
If there is any way to clear the old rules, I could add them again. That would also help.

**Comment by victorclaessen** [comment]
I can confirm that that partially fixed the problem of the error 500 that made the rules invisible; the rules are visible again.

But all the actions have disappeared from the rules. When I try to re-add an action to a rule, I get a new error 500 `Bad Arguments: {resource_not_initialized,<<"resource:c9599e60">>}` and then the rule is not saved.

Unfortunately I do not see anything in the error log on any of the nodes.

**Comment by spring2maz** [comment]
I will look into this new error.

**Comment by spring2maz** [comment]
Hi again @victorclaessen This seem to be caused by broken in-memory data consistency.
When the node boots up, it loads up the rules stored in database, and build a in-mem cache from it.
The new version node loaded up the database, but did not build the cache (due to old format not recognizable).

To fix it, please try to restart one of the nodes in the cluster,
when it starts up, it should also trigger other nodes to re-build the cache.

WRT why the cache was not built when v4.1.3 is started on v4.0.3 data:
old version rules were silently dropped when building the cache,
this is arguably a bug, because otherwise (e.g. crash) the issue would have exposed earlier.

**Comment by victorclaessen** [comment]
Ok, progress again. After restarting one of the nodes, now I get this error when I try to add an existing webhook resource back onto an existing rule.

`Bad Arguments: {clear_actions_failure,[{badrpc,{'EXIT',{function_clause,[{emqx_rule_engine,'-clear_actions/1-fun-0-',[{action_instance,<<"data_to_webserver_1587894541886874623">>,data_to_webserver,#{<<"$resource">> => <<"resource:d37db91e">>,<<"payload_tmpl">> => <<"{\"text\": \"_${topic}_: `${payload}`\"}">>}}],[{file,"/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine.erl"},{line,467}]},{lists,foreach,2,[{file,"lists.erl"},{line,1338}]},{rpc,'-handle_call_call/6-fun-0-',5,[{file,"rpc.erl"},{line,197}]}]}}},{badrpc,{'EXIT',{function_clause,[{emqx_rule_engine,'-clear_actions/1-fun-0-',[{action_instance,<<"data_to_webserver_1587894541886874623">>,data_to_webserver,#{<<"$resource">> => <<"resource:d37db91e">>,<<"payload_tmpl">> => <<"{\"text\": \"_${topic}_: `${payload}`\"}">>}}],[{file,"/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine.erl"},{line,467}]},{lists,foreach,2,[{file,"lists.erl"},{line,1338}]},{rpc,'-handle_call_call/6-fun-0-',5,[{file,"rpc.erl"},{line,197}]}]}}},{badrpc,{'EXIT',{function_clause,[{emqx_rule_engine,'-clear_actions/1-fun-0-',[{action_instance,<<"data_to_webserver_1587894541886874623">>,data_to_webserver,#{<<"$resource">> => <<"resource:d37db91e">>,<<"payload_tmpl">> => <<"{\"text\": \"_${topic}_: `${payload}`\"}">>}}],[{file,"/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine.erl"},{line,467}]},{lists,foreach,2,[{file,"lists.erl"},{line,1338}]},{rpc,'-handle_call_call/6-fun-0-',5,[{file,"rpc.erl"},{line,197}]}]}}},{badrpc,{'EXIT',{function_clause,[{emqx_rule_engine,'-clear_actions/1-fun-0-',[{action_instance,<<"data_to_webserver_1587894541886874623">>,data_to_webserver,#{<<"$resource">> => <<"resource:d37db91e">>,<<"payload_tmpl">> => <<"{\"text\": \"_${topic}_: `${payload}`\"}">>}}],[{file,"/emqx-rel/_build/emqx-pkg/lib/emqx_rule_engine/src/emqx_rule_engine.erl"},{line,467}]},{lists,foreach,2,[{file,"lists.erl"},{line,1338}]},{rpc,'-handle_call_call/6-fun-0-',5,[{file,"rpc.erl"},{line,197}]}]}}}]}`

**Comment by terry-xiaoyu** [comment]
@victorclaessen The resource has to be initialized (established) before creating the rule. That's why `Arguments: {resource_not_initialized,<<"resource:c9599e60">>}` occurs.

**Comment by victorclaessen** [comment]
@terry-xiaoyu Can you tell me how to initialize/establish the resource?

In the resources view, when I click on the 'Status' button of the resource, it shows that on each of the nodes, this resource is 'In Service', with a green dot in front.

**Comment by spring2maz** [comment]
@victorclaessen this seem to be the same cause as the in-mem cache consistency issue I mentioned before.
the `badrpc` error suggests that this error is happening in a remote node.
I must be wrong about restarting one node would make other nodes refresh the cache.
To be one the safe side, please try a rolling restart of other nodes too.

**Comment by victorclaessen** [comment]
Ok, I restarted all the nodes and the problem did not go away. I am, however, able to recreate a new rule using that resource. So my immediate problem is resolved.

But whenever I try to re-add the resource to the old rule, I get the same error 500. I think there is something wrong with the existing rule. When I try to delete the old rule, I get...you've guessed it: error 500.

**Comment by spring2maz** [comment]
I'll reproduce this error to find a fix

**Comment by spring2maz** [comment]
hi @victorclaessen 

When I tried to reproduce this issue, I found that there was a typo made when I paste my code in the reply,
it should cause a "Bad arity" exception and fail to update.
and the code in shell does not work with clustered nodes.
Not sure why it worked for you.

I need to understand the state of your database before I can provide another fix.
could you execute `rp(mnesia:table_info(emqx_rule,attributes)).` in Erlang shell, and provide the output ?
execut `emqx attach` to attach to Erlang shell.

**Comment by victorclaessen** [comment]
Thanks for staying on this! The output is:
```
[id,for,rawsql,is_foreach,fields,doeach,incase,conditions,
 on_action_failed,actions,enabled,description]
ok
```

**Comment by spring2maz** [comment]
Hi @victorclaessen 

Please find the steps below:

1. Save this code snippet as a file named `/tmp/transform_4_1_3.erl` (or some other path) to all emqx nodes in the cluster.
```
-module(transform_4_1_3).
-export([fix_rule/1, run/0]).

fix_action(A) when size(A) =:= 4 ->
  {Fields1, Fields2} = lists:split(3,tuple_to_list(A)),
  list_to_tuple(Fields1 ++ [[] | Fields2]);
fix_action(A) when size(A) =:= 5 -> A. % already fixed

fix_rule_size(Rule) when size(Rule) =:= 12 ->
  {Fields1, Fields2} = lists:split(9,tuple_to_list(Rule)),
  list_to_tuple(Fields1 ++ [continue | Fields2]);
fix_rule_size(Rule) when size(Rule) =:= 13 -> Rule. % already fixed

fix_rule(Rule) ->
  %% ensure new rule record has 13 fields (tag included)
  R1 = fix_rule_size(Rule),
  %% ensure 'on_action_failed' field is an atom (stop/continue)
  R2 = case is_atom(element(10, R1)) of
         true  -> R1;
         false -> setelement(10, R1, continue)
       end,
  %% ensure 'actions' is a list and all elements have szie=5
  NewActions = lists:map(fun fix_action/1, element(11, R2)),
  R3 = setelement(11, R2, NewActions),
  %% ensure 'enabled' field is a boolean, fix to 'true' if not
  case element(12, R3) of
    E when is_boolean(E) -> R3;
    _ -> setelement(12, true, R3)
  end.

run() ->
  mnesia:transform_table(emqx_rule,
                         fun transform_4_1_3:fix_rule/1,
                         [id,for,rawsql,is_foreach,
                          fields, doeach, incase, conditions, on_action_failed,
                          actions,enabled, description]).
```
2. Execute `c:c("/tmp/transform_4_1_3.erl").` in Erlang shell in all emqx nodes in the cluster.
3. Execute `transform_4_1_3:run().` in Erlang shell in all emqx nodes in the cluster (should be enough to execute in one node, however to be on the safe side, please do it in all nodes). This should print `{atomic, ok}`, if it doesn't please report back the errors.
4. Apply a rolling restart of the nodes in the cluster (to ensure cache reload).

**Comment by victorclaessen** [comment]
That worked perfectly. I could now delete the old rules. All is well. 

Thanks a lot for your persistence!

**Comment by spring2maz** [comment]
So glad that it’s finally fixed.
Thanks for your patience with us @victorclaessen

**Comment by docaustyne** [comment]
Having same error in 
Installed versions:
* 4.3.4	permanent
Error logs are clean and shows everything is okay, what do i do please?

**Comment by k32** [comment]
Hello @docaustyne ,
Could you reproduce the problem again and attach the latest `logs/emqx.log` file, please?

**Comment by docaustyne** [comment]
Thank you for the feedback, I somehow fixed it, version 4.3.4 persisted after an update to 4.3.5. Restarted server and I am good to go.
